src/transaction/requests.rs

```diff
@@ -160,3 +179,141 @@ impl KvRequest for kvrpcpb::ScanRequest {
+    type RpcResponse = kvrpcpb::ResolveLockResponse;
+    type KeyData = (kvrpcpb::Context, Vec<Vec<u8>>);
+    const REQUEST_NAME: &'static str = "kv_resolve_lock";
+    const RPC_FN: RpcFnType<Self, Self::RpcResponse> = TikvClient::kv_resolve_lock_async_opt;
+
+    fn make_rpc_request<KvC: KvClient>(
+        &self,
+        (context, keys): Self::KeyData,
+        _store: &Store<KvC>,
+    ) -> Self {
+        let mut req = Self::default();
+        req.set_context(context);
+        req.set_start_version(self.start_version);
+        req.set_commit_version(self.commit_version);
+        req.set_txn_infos(self.txn_infos.clone());
+        req.set_keys(keys);
+
+        req
+    }
+
+    fn on_region_error(
+        self,
+        region_error: Error,
+        _pd_client: Arc<impl PdClient>,
+    ) -> BoxStream<'static, Result<Self::RpcResponse>> {
+        stream::once(future::err(region_error)).boxed()
+    }
+
+    fn store_stream<PdC: PdClient>(
+        &mut self,
+        pd_client: Arc<PdC>,
+    ) -> BoxStream<'static, Result<(Self::KeyData, Store<PdC::KvClient>)>> {
+        let context = self
+            .context
+            .take()
+            .expect("ResolveLockRequest context must be given ");
+        let keys = mem::replace(&mut self.keys, Vec::default());
+        pd_client
+            .store_for_id(context.region_id)
+            .map_ok(move |store| ((context, keys), store))
+            .into_stream()
+            .boxed()
+    }
+
+    fn map_result(_: Self::RpcResponse) -> Self::Result {}
+
+    fn reduce(
+        results: BoxStream<'static, Result<Self::Result>>,
+    ) -> BoxFuture<'static, Result<Self::Result>> {
+        results.try_collect().boxed()
+    }
+}
+
+pub fn new_resolve_lock_request(
+    context: kvrpcpb::Context,
+    start_version: u64,
+    commit_version: u64,
+) -> kvrpcpb::ResolveLockRequest {
+    let mut req = kvrpcpb::ResolveLockRequest::default();
+    req.set_context(context);
+    req.set_start_version(start_version);
+    req.set_commit_version(commit_version);
+
+    req
+}
+
+// TODO: Add lite resolve lock (resolve specified locks only)
+
+impl KvRequest for kvrpcpb::CleanupRequest {
+    /// Commit version if the key is committed, 0 otherwise.
+    type Result = u64;
+    type RpcResponse = kvrpcpb::CleanupResponse;
+    type KeyData = Key;
+    const REQUEST_NAME: &'static str = "kv_cleanup";
+    const RPC_FN: RpcFnType<Self, Self::RpcResponse> = TikvClient::kv_cleanup_async_opt;
+
+    fn make_rpc_request<KvC: KvClient>(&self, key: Self::KeyData, store: &Store<KvC>) -> Self {
+        let mut req = store.request::<Self>();
+        req.set_key(key.into());
+        req.set_start_version(self.start_version);
+
+        req
+    }
+
+    fn store_stream<PdC: PdClient>(
+        &mut self,
+        pd_client: Arc<PdC>,
+    ) -> BoxStream<'static, Result<(Self::KeyData, Store<PdC::KvClient>)>> {
+        let key = mem::replace(&mut self.key, Default::default()).into();
```

Comment:
It seems repeated with L36-L41, could we avoid it?

Reply:
Done